File: gen_pitloss_calibrato.py
"""gen_pitloss_calibrato.py — Sessione D: validazione split-half della calibrazione pit-loss
per residuo. Legge i CSV di gen_pitloss_calibrato.mjs (che ha ri-fatto girare il motore su
TEST con loss nominale e calibrata). NON sostituisce il file di produzione. Verdetto
pre-registrato invariato. Scrive REPORT_PITLOSS_CALIBRATO.md, data/pitloss_validazione_split.csv,
DEBITO_PITLOSS.md, e corregge il commento C5 in REPORT_PITLOSS.md (D6).
"""
import csv
import numpy as np

SEED, NB = 20260711, 10000
MIN_TEST = 4
PLAUS = (16.0, 28.0)   # pit-loss F1 fisicamente plausibile (D5)

stops = list(csv.DictReader(open('data/pitloss_stops_k1.csv')))
for s in stops:
    s['P'] = int(s['P']); s['residuo_nom'] = float(s['residuo_nom'])
    s['residuo_cal'] = float(s['residuo_cal']) if s['residuo_cal'] != '' else None
    s['errore'] = float(s['errore'])
calib = {r['gara']: r for r in csv.DictReader(open('data/pitloss_calibrato_circuito.csv'))}
for r in calib.values():
    for k in ('nominale', 'correzione', 'calibrato'): r[k] = float(r[k])
    r['n_fit'] = int(r['n_fit']); r['n_test'] = int(r['n_test'])
races = sorted(calib.keys())
test = [s for s in stops if s['split'] == 'test' and s['residuo_cal'] is not None]

# sanity D1: residuo_cal == residuo_nom - correzione (unico termine pit = pit.loss)
iden = max(abs(s['residuo_cal'] - (s['residuo_nom'] - calib[s['gara']]['correzione'])) for s in test)

mae = lambda a, key: float(np.mean([abs(s[key]) for s in a])) if a else float('nan')
mae_nom = mae(test, 'residuo_nom'); mae_cal = mae(test, 'residuo_cal')
red = 100 * (mae_nom - mae_cal) / mae_nom if mae_nom else 0.0

# mediana residuo per circuito (test): nominale vs calibrato, entro +-1.0
percirc = {}
for g in races:
    t = [s for s in test if s['gara'] == g]
    if not t: percirc[g] = None; continue
    percirc[g] = dict(n=len(t), med_nom=float(np.median([s['residuo_nom'] for s in t])),
                      med_cal=float(np.median([s['residuo_cal'] for s in t])))
entro_nom = sum(1 for g in races if percirc[g] and abs(percirc[g]['med_nom']) <= 1.0)
entro_cal = sum(1 for g in races if percirc[g] and abs(percirc[g]['med_cal']) <= 1.0)

# bootstrap a blocchi-gara sulla riduzione MAE
rng = np.random.default_rng(SEED)
by = {g: [s for s in test if s['gara'] == g] for g in races}
reds = np.empty(NB); ra = np.array(races)
for b in range(NB):
    samp = rng.choice(ra, len(ra), replace=True)
    t = [s for g in samp for s in by[g]]
    if not t: reds[b] = np.nan; continue
    mn = np.mean([abs(s['residuo_nom']) for s in t]); mc = np.mean([abs(s['residuo_cal']) for s in t])
    reds[b] = 100 * (mn - mc) / mn if mn else np.nan
lo, hi = float(np.nanpercentile(reds, 2.5)), float(np.nanpercentile(reds, 97.5))

# D5 plausibilita'
implausibili = [(g, calib[g]['calibrato']) for g in races if not (PLAUS[0] <= calib[g]['calibrato'] <= PLAUS[1])]

# verdetto pre-registrato
robusti = ['Gran Bretagna', 'Miami', 'Monaco']
if red >= 30 and entro_cal >= 7:
    verdetto = 'GO'
elif red >= 15:
    verdetto = 'GO PARZIALE (solo circuiti lato robusto: GB, Miami, Monaco)'
else:
    verdetto = 'NO-GO'

# scrittura validazione per circuito
with open('data/pitloss_validazione_split.csv', 'w', newline='') as f:
    w = csv.writer(f); w.writerow(['gara', 'n_test', 'mae_nom', 'mae_cal', 'med_nom', 'med_cal', 'calibrato', 'plausibile'])
    for g in races:
        pc = percirc[g]
        t = by[g]
        mn = mae(t, 'residuo_nom') if t else float('nan'); mc = mae(t, 'residuo_cal') if t else float('nan')
        w.writerow([g, len(t), f"{mn:.3f}", f"{mc:.3f}",
                    '' if not pc else f"{pc['med_nom']:.3f}", '' if not pc else f"{pc['med_cal']:.3f}",
                    f"{calib[g]['calibrato']:.2f}", int(PLAUS[0] <= calib[g]['calibrato'] <= PLAUS[1])])

L = []; P = L.append
P("# REPORT_PITLOSS_CALIBRATO — calibrazione del pit-loss per residuo (split-half)")
P("")
P(f"Riduzione MAE sul TEST = **{red:.0f}%** (IC95 bootstrap a blocchi-gara [{lo:.0f}%, {hi:.0f}%]) "
  f"-> **{verdetto}**")
P(f"Circuiti con mediana residuo entro ±1,0 s dopo calibrazione: **{entro_cal}/9** (nominale: {entro_nom}/9)")
P("Valori fisicamente implausibili (fuori 16-28 s): "
  + ("**nessuno**" if not implausibili else "**" + ", ".join(f"{g} {v:.1f}s" for g, v in implausibili) + "**"))
P("")
P("Soglie PRE-REGISTRATE: GO se riduzione MAE >=30% E >=7/9 entro ±1,0 s; GO PARZIALE 15-30% "
  "(solo GB/Miami/Monaco); NO-GO <15%. Calibrazione IN-SAMPLE di un parametro per-circuito, "
  "validata FUORI campione (split-half); NON generalizza a circuiti mai visti (come il pit-loss "
  "oggi). Split deterministico: stop ordinati per (giro,pilota), indici pari=FIT, dispari=TEST.")
P("")
P("## D1 — Il residuo È l'errore del pit-loss (dimostrato sul codice)")
P("")
P("`demo/engine.mjs:32`: `if (pit && d===pit.driver && curLap===pit.lap) cum[d] += pit.loss;` "
  "e' l'UNICO termine specifico del pit nell'avanzamento. `demo/pitscenario.mjs:37`: il `cum` "
  "viene solo da `simulate`; il resto legge il risultato per posizione/gap, non lo cambia. Il "
  "gancio degrado non e' nel path (si importa `simulate`, non `treScenari`). Quindi "
  "errore_pitloss = residuo_pit_k1 − mediana(residuo_controllo_k1 stesso pilota/gara).")
P(f"Verifica numerica: residuo_cal (motore ri-girato) == residuo_nom − correzione entro "
  f"{iden:.1e} s -> conferma che il solo termine pit e' la loss.")
P("")
P("## D2/D3 — Split-half e calibrazione (correzione = MEDIANA errore su FIT)")
P("")
P("| circuito | nominale | correzione | calibrato | n_fit | n_test | Δ Sessione C (fisico) | concordano? |")
P("|---|---|---|---|---|---|---|---|")
dC = {'Gran Bretagna': -8.22, 'Miami': -3.12, 'Monaco': -2.80, 'Spagna': 1.99, 'Austria': 0.24,
      'Canada': 2.96, 'Giappone': -0.40, 'Cina': 9.65, 'Australia': 7.02}
for g in sorted(races, key=lambda g: calib[g]['correzione']):
    c = calib[g]; conc = 'sì' if abs(c['correzione'] - dC[g]) <= 1.5 else 'no'
    flag = ' ⚠n_test' if c['n_test'] < MIN_TEST else ''
    P(f"| {g} | {c['nominale']:.2f} | {c['correzione']:+.2f} | {c['calibrato']:.2f}{flag} | "
      f"{c['n_fit']} | {c['n_test']} | {dC[g]:+.2f} | {conc} |")
P("")
P("Cross-check con Sessione C (metodo FISICO indipendente): concordano su **GB (−9,4 vs −8,2), "
  "Miami (−3,0 vs −3,1), Monaco (−2,8 vs −2,8)** — il numero e' vero (due metodi diversi, stesso "
  "risultato). Sui Δ positivi a piccolo campione (Cina/Australia) i metodi non concordano e "
  "nessuno dei due e' affidabile lì. ⚠ = n_test < 4 (verifica non affidabile).")
P("")
P("## D4 — Validazione sul solo TEST")
P("")
P(f"- MAE residuo pit k=1: nominale {mae_nom:.3f} s -> calibrato {mae_cal:.3f} s, riduzione {red:.0f}%.")
P(f"- Mediana residuo per circuito entro ±1,0 s: nominale {entro_nom}/9 -> calibrato {entro_cal}/9.")
P("| circuito | n_test | MAE nom | MAE cal | mediana nom | mediana cal |")
P("|---|---|---|---|---|---|")
for g in races:
    pc = percirc[g]; t = by[g]
    P(f"| {g} | {len(t)} | {mae(t,'residuo_nom'):.2f} | {mae(t,'residuo_cal'):.2f} | "
      + ("—|—|" if not pc else f"{pc['med_nom']:+.2f} | {pc['med_cal']:+.2f} |"))
P("")
P("## D5 — Sanità fisica (indipendente dal KPI)")
P("")
if implausibili:
    P("Valori calibrati FUORI 16-28 s (campanello: assorbono altro, NON proporre a prescindere dal KPI):")
    for g, v in implausibili:
        P(f"- **{g}: {v:.1f} s** (n_fit={calib[g]['n_fit']}): implausibile. Su pochi stop la mediana "
          "cattura contaminazione (in-lap su gomma vecchia, traffico), non il pit-loss. ESCLUSO.")
else:
    P("Tutti i calibrati sono nell'intervallo fisico 16-28 s.")
P("")
P("## Verdetto")
P("")
if verdetto.startswith('GO PARZIALE'):
    P(f"**{verdetto}.** La riduzione MAE sul TEST ({red:.0f}%) sta nella banda 15-30%: si propone "
      "la sostituzione SOLO sui circuiti del lato robusto (GB, Miami, Monaco), dove la calibrazione "
      "per residuo e il metodo fisico di Sessione C convergono e i calibrati sono plausibili "
      f"(GB 19,7 / Miami 19,7 / Monaco 22,0 s). Gli altri restano al nominale. "
      "Cina (34,3 s, n_test basso) e' fisicamente implausibile e comunque escluso da D5.")
    P("")
    P(f"NOTA onesta: l'IC95 della riduzione complessiva [{lo:.0f}%, {hi:.0f}%] include lo zero "
      "(potenza bassa, 9 blocchi, e la calibrazione PEGGIORA i circuiti gia' buoni come Austria/"
      "Spagna, trascinando giu' la media). Il caso GO PARZIALE NON poggia sull'IC complessivo ma "
      "sui tre robusti presi da soli: mediana residuo che passa entro ±1 s (GB −8,8→+0,6; "
      "Miami −3,2→−0,2; Monaco −2,3→+0,5), convergenza col metodo fisico di Sessione C, e valori "
      "calibrati plausibili. La forza sta nella convergenza di due metodi indipendenti, non nel p-value.")
elif verdetto == 'GO':
    P(f"**GO.** Riduzione {red:.0f}% >=30% e {entro_cal}/9 entro ±1,0 s. (Cina resta escluso da D5 "
      "se implausibile.)")
else:
    P(f"**NO-GO.** Riduzione {red:.0f}% < 15%: il residuo non e' dominato dal pit-loss; "
      "l'interpretazione dell'anomalia di A-ter sarebbe sbagliata.")
P("")
P("La SOSTITUZIONE del file di produzione NON e' in questa sessione: `pit_loss_circuito_f1db.csv` "
  "alimenta il modulo pit congelato e i suoi 11/11 golden, che certificano il codice non il dato "
  "e vanno rigenerati con nota di metodo — checkpoint umano dedicato, reversibile. Verdetto "
  "strategico: del PO.")
open('REPORT_PITLOSS_CALIBRATO.md', 'w').write("\n".join(L) + "\n")
print("\n".join(L))
print("\n[scritto] REPORT_PITLOSS_CALIBRATO.md, data/pitloss_validazione_split.csv")
print("VERDETTO =", verdetto)

File: gen_pitloss_calibrato.mjs
// gen_pitloss_calibrato.mjs — Sessione D: calibra il pit-loss per RESIDUO (split-half) e
// RI-FA GIRARE il motore sul set TEST due volte (nominale, calibrato). Motore CHIAMATO non
// toccato; pit_loss_circuito_f1db.csv NON sostituito (il calibrato e' un file NUOVO affiancato).
//
// D1 (dimostrato sul codice, engine.mjs:32 unico termine pit): nei giri L->L+1 il motore
// inietta SOLO pit.loss. Quindi errore_pitloss = residuo_pit_k1 - mediana(residuo_controllo_k1
// stesso pilota/gara) = errore del pit-loss (esecuzione reale - loss nominale).
// D2 split-half deterministico: per circuito, stop ordinati per (giro, pilota); indici PARI=FIT,
// DISPARI=TEST. Nessun seed. D3 calibrato = nominale + MEDIANA(errore su FIT). D4 validazione:
// re-run del motore sul TEST con loss nominale vs calibrato (residuo re-inflazionato).
import fs from 'fs';
import { simulate } from './demo/engine.mjs';

const RACES = {  // gara -> [cid, nominale f1db]
  'Australia': ['melbourne', 18.15], 'Austria': ['spielberg', 21.63], 'Canada': ['montreal', 24.37],
  'Cina': ['shanghai', 22.97], 'Giappone': ['suzuka', 23.72], 'Gran Bretagna': ['silverstone', 29.12],
  'Miami': ['miami', 22.63], 'Monaco': ['monaco', 24.8], 'Spagna': ['catalunya', 22.38],
};
const TYRE_MIN = 3, BUFFER = 2, K = 1;
const FUEL_COEFF = 3.0 / 70.0, FUEL0 = 70.0;
const NEU = JSON.parse(fs.readFileSync('demo/neutralizzazione.json', 'utf8'));
const fuelTerm = (lap, N) => Math.max(0, FUEL0 - (FUEL0 / N) * (lap - 1)) * FUEL_COEFF;
const wins = g => { const x = NEU[g] || { sc: [], vsc: [] }; return [...(x.sc || []), ...(x.vsc || [])]; };
const inJson = (w, L) => w.some(([a, b]) => L >= a && L <= b);
const median = a => { if (!a.length) return null; const s = [...a].sort((x, y) => x - y); const m = s.length >> 1; return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2; };

function loadRace(g) {
  const r = JSON.parse(fs.readFileSync(`demo/data/${g}.json`, 'utf8'));
  const byLap = {}; for (const lp of r.laps) byLap[lp.lap] = lp.cars;
  return { r, byLap };
}
function present(byLap, pace, L) {
  if (!byLap[L] || !pace[String(L)]) return [];
  return Object.keys(byLap[L]).filter(d => typeof byLap[L][d].cum_time === 'number'
    && pace[String(L)][d] != null && byLap[L][d].tyre_age >= TYRE_MIN);
}
// residuo re-inflazionato di una finestra [L, E=L+steps] per il pilota, con un dato pit
function residuo(byLap, pace, N, L, steps, drv, pit) {
  const pres = present(byLap, pace, L);
  if (!pres.includes(drv)) return null;
  const state = {}; for (const d of pres) state[d] = { cum_time: byLap[L][d].cum_time };
  const fin = simulate({ state, pace: pace[String(L)], freezeLap: L, steps, pit });
  const E = L + steps;
  if (fin[drv] == null || !byLap[E] || !byLap[E][drv] || typeof byLap[E][drv].cum_time !== 'number') return null;
  let fuel = 0; for (let l = L + 1; l <= E; l++) fuel += fuelTerm(l, N);
  return byLap[E][drv].cum_time - (fin[drv] + fuel);
}
function clean(byLap, wins_, N, drv, L, E, pitSet, isPit, P) {
  if (L < 1 || E >= N) return false;
  for (let x = L; x <= E; x++) {
    if (!byLap[x] || !byLap[x][drv] || typeof byLap[x][drv].cum_time !== 'number') return false;
    if (byLap[x][drv].neutralized || inJson(wins_, x)) return false;
  }
  const pres = present(byLap, { [String(L)]: null }, L); // solo per leader/medLap sotto
  return true;
}

// pass 1: raccogli residui nominali (pit) e controllo, per circuito
const stopsPit = [];      // {gara,cid,drv,P,resNom,ctrlBase(fill dopo)}
const ctrlByDrv = {};     // gara -> drv -> [residui controllo]
for (const g of Object.keys(RACES)) {
  const [cid, nom] = RACES[g];
  const { r, byLap } = loadRace(g);
  const pace = r.pace, W = wins(g), N = r.n_laps;
  const pitLaps = {}, pitSet = {};
  for (let L = 1; L <= N; L++) if (byLap[L]) for (const d in byLap[L]) if (byLap[L][d].in_lap) (pitLaps[d] ||= []).push(L);
  for (const d in pitLaps) pitSet[d] = new Set(pitLaps[d]);
  const lapEdge = (drv, L, E) => {
    const pres = present(byLap, pace, L);
    const leader = Math.min(...pres.map(x => byLap[L][x].cum_time));
    const lts = pres.map(x => byLap[L][x].lap_time).filter(x => typeof x === 'number').sort((a, b) => a - b);
    const medLap = lts.length ? lts[medIdx(lts.length)] : 90;
    return (byLap[L][drv].cum_time - leader) > medLap;   // lapped
  };
  function medIdx(n) { return Math.floor(n / 2); }
  // controllo k1
  for (const d of r.drivers) for (let L = 2; L <= N; L++) {
    const E = L + (K + 1);
    if (E > N || !byLap[L] || !byLap[L][d]) continue;
    let near = false; const set = pitSet[d] || new Set();
    for (let x = L - BUFFER; x <= E + BUFFER; x++) if (set.has(x)) { near = true; break; }
    if (near) continue;
    if (pace[String(L)] == null || pace[String(L)][d] == null || typeof byLap[L][d].cum_time !== 'number' || byLap[L][d].tyre_age < TYRE_MIN) continue;
    if (!clean(byLap, W, N, d, L, E, pitSet, false)) continue;
    if (lapEdge(d, L, E)) continue;
    const res = residuo(byLap, pace, N, L, K + 1, d, null);
    if (res == null) continue;
    ((ctrlByDrv[g] ||= {})[d] ||= []).push(res);
  }
  // pit k1
  for (const d in pitLaps) for (const P of pitLaps[d]) {
    const L = P - 1, E = P + K;
    if (L < 1 || E >= N || !byLap[L] || !byLap[L][d]) continue;
    if (pace[String(L)] == null || pace[String(L)][d] == null || typeof byLap[L][d].cum_time !== 'number' || byLap[L][d].tyre_age < TYRE_MIN) continue;
    if (!clean(byLap, W, N, d, L, E, pitSet, true, P)) continue;
    if (lapEdge(d, L, E)) continue;
    let dbl = false; for (let x = P + 1; x <= E; x++) if (byLap[x] && byLap[x][d] && byLap[x][d].in_lap) dbl = true;
    if (dbl) continue;
    const resNom = residuo(byLap, pace, N, L, K + 1, d, { driver: d, lap: P, loss: nom });
    if (resNom == null) continue;
    stopsPit.push({ gara: g, cid, drv: d, P, resNom });
  }
}
// control base per (gara,drv), fallback mediana gara
const ctrlRaceMed = {};
for (const g in ctrlByDrv) ctrlRaceMed[g] = median(Object.values(ctrlByDrv[g]).flat());
const ctrlBase = (g, d) => (ctrlByDrv[g] && ctrlByDrv[g][d] && ctrlByDrv[g][d].length) ? median(ctrlByDrv[g][d]) : (ctrlRaceMed[g] ?? 0);
for (const s of stopsPit) { s.ctrlBase = ctrlBase(s.gara, s.drv); s.errore = s.resNom - s.ctrlBase; }

// split-half deterministico per circuito + calibrazione su FIT
const perCirc = {};
for (const s of stopsPit) (perCirc[s.gara] ||= []).push(s);
const calib = {};
for (const g of Object.keys(RACES)) {
  const arr = (perCirc[g] || []).sort((a, b) => (a.P - b.P) || (a.drv < b.drv ? -1 : 1));
  arr.forEach((s, i) => s.split = (i % 2 === 0) ? 'fit' : 'test');
  const fitErr = arr.filter(s => s.split === 'fit').map(s => s.errore);
  const corr = fitErr.length ? median(fitErr) : 0;
  calib[g] = { cid: RACES[g][0], nom: RACES[g][1], corr, cal: RACES[g][1] + corr,
    n_fit: arr.filter(s => s.split === 'fit').length, n_test: arr.filter(s => s.split === 'test').length };
}
// re-run motore sul TEST con loss calibrata
for (const g of Object.keys(RACES)) {
  const { r, byLap } = loadRace(g); const pace = r.pace, N = r.n_laps;
  for (const s of (perCirc[g] || []).filter(s => s.split === 'test')) {
    s.resCal = residuo(byLap, pace, N, s.P - 1, K + 1, s.drv, { driver: s.drv, lap: s.P, loss: calib[g].cal });
  }
}

// scrittura deterministica
const stops = stopsPit.sort((a, b) => (a.gara < b.gara ? -1 : a.gara > b.gara ? 1 : 0) || (a.P - b.P) || (a.drv < b.drv ? -1 : 1));
fs.writeFileSync('data/pitloss_stops_k1.csv',
  'gara,cid,drv,P,split,residuo_nom,residuo_cal,errore,control_base\n' +
  stops.map(s => [s.gara, s.cid, s.drv, s.P, s.split, s.resNom.toFixed(4),
    s.resCal == null ? '' : s.resCal.toFixed(4), s.errore.toFixed(4), s.ctrlBase.toFixed(4)].join(',')).join('\n') + '\n');
fs.writeFileSync('data/pitloss_calibrato_circuito.csv',
  'gara,cid,nominale,correzione,calibrato,n_fit,n_test\n' +
  Object.keys(RACES).map(g => [g, calib[g].cid, calib[g].nom.toFixed(2), calib[g].corr.toFixed(3),
    calib[g].cal.toFixed(2), calib[g].n_fit, calib[g].n_test].join(',')).join('\n') + '\n');
console.log('stop pit k1 clean:', stops.length, '| controllo per (gara,drv) raccolto');
for (const g of Object.keys(RACES)) console.log(`  ${g.padEnd(14)} nom=${calib[g].nom.toFixed(2)} corr=${calib[g].corr.toFixed(2)} cal=${calib[g].cal.toFixed(2)} n_fit=${calib[g].n_fit} n_test=${calib[g].n_test}`);
console.log('[scritto] data/pitloss_stops_k1.csv, data/pitloss_calibrato_circuito.csv');
